### PR TITLE
Allow "compiling" a course to a directory

### PR DESCRIPTION
**Pull request by encukou**

### CODE REVIEW

**Review by hroncok — COMMENTED** [review]
This is a big chunk of diff and I haven't tested anything, but overall the changes make sense.

Added some inline questions that would help me understand things better or nit-pick suggestions.

File: test_naucse_render/test_integration.py [hunk 1]
```diff
@@ -14,11 +14,17 @@
         '2000/run-with-times',
         '2000/run-with-timezone',
         'lessons',
+        '',
```
[the remark this answers is not in the document]
Comment (encukou) [reply test_naucse_render/test_integration.py]: I can make it two options, missing and `None`.

[the remark this answers is not in the document]
Comment (hroncok) [reply test_naucse_render/test_integration.py]: I mean, what is the differnce?

File: naucse_render/compile.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    lesson_slugs = get_lesson_slugs(course_info)
+    vars = course_info.get('vars')
+    response = get_lessons(lesson_slugs, path=path, vars=vars)
+    course_info['lessons'] = response['data']
+
+    info_path = destination / 'course.json'
+    if destination.exists():
+        if (
+            not info_path.exists()
+            and any(info_path.glob('*'))
```
[the remark this answers is not in the document]
Comment (encukou) [reply naucse_render/compile.py:25]: Not more straightforward, but I imagine it's faster. Thanks!

File: naucse_render/compile.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        for material_info in session_info.get('materials', ()):
+            lesson_slug = material_info.get('lesson_slug')
+            if lesson_slug:
+                slugs.add(lesson_slug)
+    return sorted(slugs)
+
+
+def unique_path(path):
+    """Generate an unused filename that looks like `path`"""
+    # this adds ".1", ".2" etc. before the extension
```
[the remark this answers is not in the document]
Comment (encukou) [reply naucse_render/compile.py:57]: I didn't want to copy the directory structure of the source repo, and I don't want to worry about conflicts filenames when I make future changes (e.g. allow courses made from several sources of lessons.)

So the names and directory structure are meaningless, they just need to be unique (except `course.json`, which references the other files by name). I could just use a hash of the contents for all filenames, but that would make it harder for humans to inspect the result, and probably hurt Git's compression heuristics.

I deliberately put in a possibility to get a name conflict, so I can test this. (Specifically: the lesson category is omitted, so `beginners/install-editor` and `duplicate/install-editor` lessons in the default test course overlap.)

I'll expand the docstring.

File: test_naucse_render/test_unique_path.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+def test_unique_path(tmp_path):
+    path = tmp_path / 'file.ext'
+    paths = set()
+    for i in range(10):
+        assert len(paths) == i
+        paths.add(path)
+        path.write_text(str(i))
+        path = unique_path(path)
+        assert tmp_path in path.parents
```
[the remark this answers is not in the document]
Comment (encukou) [reply test_naucse_render/test_unique_path.py:11]: This is meant as a blackbox test. If the function changed `tmp_path/file.foo` to `tmp_path/1/file.foo`, it should pass.

File: test_naucse_render/test_integration.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (encukou) [reply test_naucse_render/test_integration.py]: No difference. I did the CLI first and `''` was more natural than `None` there. In the Python API `''` doesn't make much sense.

File: naucse_render/compile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (encukou) [reply naucse_render/compile.py]: (Also, this had a bug. I added tests for it.)

File: README.md [hunk 5]
```diff
@@ -74,6 +81,21 @@ licensed under the same license.
...
+* A new subcommand, `compile`, creates a directory with course data
+  and supporting files.
+
+* API version 0.3
+  * A course may have information about "edit info", specifying where the
+    course's sources can be edited. (This is only set when compiling
+    courses; in other cases the caller should know ehrer the code lives.)
+
+* The `slug` is now optional; if not given (or given as the empty
+  string), the data is loaded from `course.yml` rather than a file in
```
[the remark this answers is not in the document]
Comment (encukou) [reply README.md]: From `path`, which defaults to current directory. I added a note here.

File: naucse_render/compile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hroncok) [reply naucse_render/compile.py]: I see now. Should check if `destination` has files in it, not `info_path`.

File: naucse_render/course.py [hunk 6]
```diff
@@ -13,7 +13,7 @@
 from .encode import encode_for_json, API_VERSION
 
 
-def get_course(course_slug: str = '', *, path='.', version=None):
+def get_course(course_slug: str = None, *, path='.', version=None):
```
Comment (hroncok) [inline naucse_render/course.py:16]: I am not good at type hints, but should str be wrapped in Optional now?

Comment (encukou) [reply naucse_render/course.py:16]: That depends on the type checker. I write for humans, and consider this understandable.

**Review by hroncok — APPROVED** [review]

**Comment by encukou** [comment]
Thanks!
Before merging I'll do integration testing with this, https://github.com/pyvec/naucse/pull/52, and naucse.python.cz

**Pull request merged**